### PR TITLE
[CONTP-512] Add settings on sidecar to terminate on finished job

### PR DESCRIPTION
**Pull request by gabedos**
### What does this PR do?

Terminates the agent when running as a sidecar when there are no other processes left running in the job.

### Motivation

[Support case](https://datadoghq.atlassian.net/browse/CONS-6814) with customer having a non-terminating sidecar agent that would cause the job to never be completed.

### Describe how to test/QA your changes

Deploy the cluster agent with the sidecar injection enabled
```
datadog:
  kubelet:
    tlsVerify: false
agents:
  enabled: false
clusterAgent:
  image:
    tag: <INSERT_RC_TAG>
    pullPolicy: Always
  enabled: true
  replicas: 1
  admissionController:
    agentSidecarInjection:
      enabled: true
      provider: fargate
```

Deploy a dummy job that allows the agent sidecar to be injected via `k apply -f sidecar-job.yaml`
```
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
        - name: pi
          image: perl:5.34.0
          command: ["perl", "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
    metadata:
      labels:
        agent.datadoghq.com/sidecar: fargate
  backoffLimit: 4
```

Note: make sure you have a datadog-secret with an api-key and app-key defined wherever you deploy the job. Otherwise run the following
```
kubectl create secret generic datadog-secret -n <INSERT_NAMESPACE> \
        --from-literal api-key=$DD_API_KEY --from-literal token=random32characterstringfortoken1
```

You should expect for a pod to be deployed with 2 containers, one being the pi container and the other being the agent container. After about 15 seconds you should see only 1/2 containers are running as the pi calculation terminates. After about 60 seconds you should `0/2 Completed` for the pi pod.

### Possible Drawbacks / Trade-offs

N/A

### Additional Notes

Also works for cronJobs

### CODE REVIEW

**Review by estherk15 — APPROVED** [review]

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go [hunk 1]
```diff
@@ -177,12 +177,25 @@ func (w *Webhook) injectAgentSidecar(pod *corev1.Pod, _ string, _ dynamic.Interf
...
 		if pod.Spec.Containers[i].Name == agentSidecarContainerName {
+			if isOwnedByJob(pod.OwnerReferences) {
+				updated, err = withEnvOverrides(&pod.Spec.Containers[i], corev1.EnvVar{
+					Name:  "DD_AUTO_EXIT_NOPROCESS_ENABLED",
+					Value: "true",
+				})
+			}
+			if err != nil {
+				log.Errorf("Failed to apply env overrides: %v", err)
+				return podUpdated, errors.New(metrics.InvalidInput)
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:188]: for the error, do we have `metrics.InternalError` or something similar?

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go [hunk 2]
```diff
@@ -320,3 +333,13 @@ func labelSelectors(datadogConfig config.Component) (namespaceSelector, objectSe
 
 	return namespaceSelector, objectSelector
 }
+
+// isOwnedByJob returns true if the pod is owned by a Job
+func isOwnedByJob(ownerReferences []metav1.OwnerReference) bool {
+	for _, owner := range ownerReferences {
+		if owner.APIVersion == "batch/v1" && owner.Kind == "Job" {
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:340]: to be futurproof, I would suggest to check only the apigroup name and not the version. 

something like
```suggestion
		if strings.HasPrefix(owner.APIVersion , "batch/") && owner.Kind == "Job" {
```

**Review by clamoriniere — COMMENTED** [review]
small nits, 
otherwise looks good 👏

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go (hunk 2 above)
Comment (gabedos) [reply pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:340]: Makes a lot of sense. Thanks for pointing this out

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**